### ISSUE TITLE
Rewrite our Grad Sample tests using Hypothesis

### ISSUE THREAD

**Issue by Darktex** [pytorch/opacus#122]
## 🚀 Feature
Rewrite our [Grad Sample tests](https://github.com/pytorch/opacus/tree/master/opacus/tests/grad_samples) using [Hypothesis](https://hypothesis.readthedocs.io/en/latest/).

## Motivation

In #120 we introduce [Hypothesis](https://hypothesis.readthedocs.io/en/latest/) as our framework of choice for property-based testing and we use it to test that the API of our DP layers is an exact match for that of their "normal" PyTorch counterparts.

Hypothesis is a great way of testing complex combinations of parameters, and it is far superior to the "homebrew" version of that that we have been using for the Grad Sample tests. We should switch and remain consistent!

## Pitch
By adding Hypothesis, we can make tests that are more legible while also testing for way more stuff and being even more thorough!

## Alternatives
Keep as-is and live our lives, but we are leaving some behavior untested, and making testing future components harder :)

## Additional context

Read #119 for additional context around this - it was written with the very purpose of giving you context for this :)

Also, it's good to outline the purpose of these tests as they are different:

- Tests for DP Layers only test the layers we do reimplement, and only test their API: we check that outputs (after a forward) and param gradients (after a backwards) are an exact match. We don't compute grad_samples here, and we don't do any DP there. Just check that the reimplemented layer is indeed trustworthy as a "normal" PyTorch layer.
- Tests for GradSamples test that we can indeed compute per-sample gradients correctly in our vectorized implementation. We test these by taking a nn.Module, computing per-sample gradients via Opacus and comparing them with the per-sample gradients we would obtain going sample by sample. These tests apply to all layers we explicitly support, including those coming from PyTorch (we assume the layer itself is correct, but we want to test our code to compute grad_samples of course).

Just like with DP Layers, tests for GradSamples should also test for varying parameters: what if our method only works with even conv kernel sizes, or if we have a bug when a dimension is set to 1? In the current implementation, we don't really randomize and we just test for a few things manually, such as the dimensions of the input (see eg [linear_test.py](https://github.com/pytorch/opacus/blob/master/opacus/tests/grad_samples/linear_test.py)).

In writing these tests, I put some common helpers in [common.py](https://github.com/pytorch/opacus/blob/master/opacus/tests/grad_samples/common.py). The resulting API is a bit different from that of the common utilities I introduced in #120 and I think the latter is the better way (for example, passing `*args` and `**kwargs` together with a `train_fn` seems a better way than introducing a `ModelWithLoss` class).

**Comment by Darktex** [comment]
Done!
